### PR TITLE
Feature/81620 - Send email to constrcomp on complete

### PR DESCRIPTION
**Pull request by janhagevold**
- Add email service
- Add send email post complete event handler

When an IPO is completed by Contractor, and email should be sent to Construction Company representative to notify that the invitation is ready for review .

### CODE REVIEW

File: src/Equinor.ProCoSys.IPO.Command/EventHandlers/PostSaveEvents/IpoCompletedEmailHandler.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+using System.Threading.Tasks;
+using Equinor.ProCoSys.IPO.Domain.AggregateModels.InvitationAggregate;
+using Equinor.ProCoSys.IPO.Domain.Events.PostSave;
+using Equinor.ProCoSys.IPO.Email;
+using MediatR;
+using Microsoft.Extensions.Options;
+
+namespace Equinor.ProCoSys.IPO.Command.EventHandlers.PostSaveEvents
+{
+    public class IpoCompletedEmailHandler : INotificationHandler<IpoCompletedEvent>
```
Comment (llykkelarsen) [inline src/Equinor.ProCoSys.IPO.Command/EventHandlers/PostSaveEvents/IpoCompletedEmailHandler.cs:12]: Could this be made part of the IpoCompletedEventHandler?

File: src/Equinor.ProCoSys.IPO.Command/EventHandlers/PostSaveEvents/IpoCompletedEmailHandler.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        } 
+
+        public Task Handle(IpoCompletedEvent notification, CancellationToken cancellationToken)
+        {
+            var constructionCompanyParticipant = notification.Invitation.Participants.SingleOrDefault(p =>
+                p.SortKey == 1 && p.Organization == Organization.ConstructionCompany);
+
+            if (constructionCompanyParticipant?.Email == null)
+            {
+                return Task.CompletedTask;
```
Comment (llykkelarsen) [inline src/Equinor.ProCoSys.IPO.Command/EventHandlers/PostSaveEvents/IpoCompletedEmailHandler.cs:30]: Does this mean that if participant is a functional role and does not have a group email, no one will be notified? Should probably add a comment here as the logic is a little tricky.

**Comment by janid1967** [comment]
I suggest at least 2 reviews must approve

File: src/Equinor.ProCoSys.IPO.Command/EventHandlers/PostSaveEvents/IpoCompletedEmailHandler.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
+﻿using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Equinor.ProCoSys.IPO.Domain.AggregateModels.InvitationAggregate;
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Command/EventHandlers/PostSaveEvents/IpoCompletedEmailHandler.cs:4]: remove ref to Domain
The IpiCompetedEvent should have bits and pieces needed, not whole Inivitation object

File: src/Equinor.ProCoSys.IPO.WebApi/DiModules/ApplicationModule.cs [hunk 4]
```diff
@@ -51,6 +53,7 @@ public static void AddApplicationModules(this IServiceCollection services, IConf
             services.Configure<BlobStorageOptions>(configuration.GetSection("BlobStorage"));
             services.Configure<AuthenticatorOptions>(configuration.GetSection("Authenticator"));
             services.Configure<MeetingOptions>(configuration.GetSection("Meetings"));
+            services.Configure<EmailOptions>(configuration.GetSection("Email"));
```
Comment (paal123) [inline src/Equinor.ProCoSys.IPO.WebApi/DiModules/ApplicationModule.cs:56]: Section is missing in appsettings.json

File: src/Equinor.ProCoSys.IPO.Email/Equinor.ProCoSys.IPO.Email.csproj [hunk 5]
```diff
@@ -0,0 +1,11 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (paal123) [inline src/Equinor.ProCoSys.IPO.Email/Equinor.ProCoSys.IPO.Email.csproj:1]: I suggest not creating a separate project for this, and rather have the service in existing project structure

File: src/Equinor.ProCoSys.IPO.Command/EventHandlers/PostSaveEvents/IpoCompletedEmailHandler.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+            var subject = $"Completed notification: IPO-{id}";
+            var body =
+                $"<p>IPO-{id}: {title} has been completed and is ready for your attention to sign and accept.</p>" +
+                "<p>Click the link to review " +
+                $"<a href=\"{baseUrl}{plant}/InvitationForPunchOut/{id}\">IPO-{id}</a>.</p>";
+
+            return _emailService.SendEmailAsync(constructionCompanyParticipant.Email, subject, body, cancellationToken);
+        }
+
+        private static string GetPlantName(string plant) => plant.Contains('$') ? plant.Split('$')[1] : plant;
```
Comment (paal123) [inline src/Equinor.ProCoSys.IPO.Command/EventHandlers/PostSaveEvents/IpoCompletedEmailHandler.cs:47]: Plant will always contain $

File: src/Equinor.ProCoSys.IPO.Command/EventHandlers/PostSaveEvents/IpoCompletedEmailHandler.cs (hunk 1 above)
Comment (paal123) [inline src/Equinor.ProCoSys.IPO.Command/EventHandlers/PostSaveEvents/IpoCompletedEmailHandler.cs:12]: We should discuss how we handle this, as this is the first time we have two different "listeners" for the same event. I think we should have one handler handling the IpoCompletedEvent.  

If email service is down, we currently miss these emails, as they are not stored until handled

**Review by paal123 — CHANGES_REQUESTED** [review]

File: src/Equinor.ProCoSys.IPO.Domain/Events/PostSave/IpoCompletedEvent.cs [hunk 7]
```diff
@@ -7,12 +8,15 @@ public class IpoCompletedEvent : INotification
...
+            Guid objectGuid,
+            Invitation invitation)
         {
             Plant = plant;
             ObjectGuid = objectGuid;
+            Invitation = invitation;
         }
         public string Plant { get; }
         public Guid ObjectGuid { get; }
+        public Invitation Invitation { get; }
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Domain/Events/PostSave/IpoCompletedEvent.cs:20]: remove. Add ints and strings needed in this context

File: src/Equinor.ProCoSys.IPO.Email/EmailService.cs [hunk 8]
```diff
@@ -0,0 +1,34 @@
+﻿using System.Net;
+using System.Net.Mail;
+using System.Threading;
+using System.Threading.Tasks;
+using Equinor.ProCoSys.IPO.Email.Settings;
+using Microsoft.Extensions.Options;
+
+namespace Equinor.ProCoSys.IPO.Email
+{
+    public class EmailService : IEmailService
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Email/EmailService.cs:10]: can't see any unit tests. Add new unit-test project below Tests folder

File: src/Equinor.ProCoSys.IPO.WebApi/Equinor.ProCoSys.IPO.WebApi.csproj [hunk 9]
```diff
@@ -39,6 +39,7 @@
     <ProjectReference Include="..\Equinor.ProCoSys.IPO.BlobStorage\Equinor.ProCoSys.IPO.BlobStorage.csproj" />
     <ProjectReference Include="..\Equinor.ProCoSys.IPO.Command\Equinor.ProCoSys.IPO.Command.csproj" />
     <ProjectReference Include="..\Equinor.ProCoSys.IPO.Domain\Equinor.ProCoSys.IPO.Domain.csproj" />
+    <ProjectReference Include="..\Equinor.ProCoSys.IPO.Email\Equinor.ProCoSys.IPO.Email.csproj" />
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.WebApi/Equinor.ProCoSys.IPO.WebApi.csproj:42]: remove ref to Email

File: src/Equinor.ProCoSys.IPO.Domain/AggregateModels/InvitationAggregate/Invitation.cs [hunk 10]
```diff
@@ -197,7 +197,7 @@ public void CompleteIpo(Participant participant, string participantRowVersion, P
             CompletedBy = completedBy.Id;
             CompletedAtUtc = completedAtUtc;
             AddPreSaveDomainEvent(new IpoCompletedEvent(Plant, ObjectGuid));
-            AddPostSaveDomainEvent(new Events.PostSave.IpoCompletedEvent(Plant, ObjectGuid));
+            AddPostSaveDomainEvent(new Events.PostSave.IpoCompletedEvent(Plant, ObjectGuid, this));
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Domain/AggregateModels/InvitationAggregate/Invitation.cs:200]: Invititationtests has one test named CompleteIpo_ShouldAddCompleteIpoEvent
This should be 2 tests:
CompleteIpo_ShouldAddCompleteIpoPreSaveEvent
CompleteIpo_ShouldAddCompleteIpoPostSaveEvent

@llykkelarsen Add a TECH to verify if these tests are forgotten elsewhere (or @janhagevold fix it now all places)

**Review by janid1967 — CHANGES_REQUESTED** [review]
See comments

File: src/Equinor.ProCoSys.IPO.Command/EventHandlers/PostSaveEvents/IpoCompletedEmailHandler.cs (hunk 6 above)
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Command/EventHandlers/PostSaveEvents/IpoCompletedEmailHandler.cs:47]: GetPlantId

File: src/Equinor.ProCoSys.IPO.Command/EventHandlers/PostSaveEvents/IpoCompletedEmailHandler.cs [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+            if (constructionCompanyParticipant?.Email == null)
+            {
+                return Task.CompletedTask;
+            }
+
+            var baseUrl = _meetingOptions.CurrentValue.PcsBaseUrl;
+            var id = notification.Invitation.Id;
+            var title = notification.Invitation.Title;
+            var plant = GetPlantName(notification.Plant);
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Command/EventHandlers/PostSaveEvents/IpoCompletedEmailHandler.cs:36]: rename plant to plantId

**Comment by llykkelarsen** [comment]
Either @paal123 or @janid1967 needs to look at this before you complete the pull request, I don't feel comfortable being the only one.

File: src/Equinor.ProCoSys.IPO.Command/EventHandlers/PostSaveEvents/IpoCompletedEventHandler.cs [hunk 12]
```diff
@@ -2,19 +2,34 @@
...
+        public IpoCompletedEventHandler(IPcsBusSender pcsBusSender, IEmailService emailService, IOptionsMonitor<MeetingOptions> meetingOptions)
+        {
+            _pcsBusSender = pcsBusSender;
+            _emailService = emailService;
+            _meetingOptions = meetingOptions;
+        } 
 
         public async Task Handle(IpoCompletedEvent notification, CancellationToken cancellationToken)
+        {
+            await SendBusTopic(notification);
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Command/EventHandlers/PostSaveEvents/IpoCompletedEventHandler.cs:28]: SendBusTopicAsync

File: src/Equinor.ProCoSys.IPO.Command/EventHandlers/PostSaveEvents/IpoCompletedEventHandler.cs [hunk 13]
```diff
@@ -2,19 +2,34 @@
...
+            await SendEmail(notification, cancellationToken);
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Command/EventHandlers/PostSaveEvents/IpoCompletedEventHandler.cs:29]: SendEmailAsync

File: src/Equinor.ProCoSys.IPO.Email/EmailService.cs (hunk 8 above)
Comment (janid1967) [reply src/Equinor.ProCoSys.IPO.Email/EmailService.cs:13]: Have we discussed that this don't deserve any unit test? Was it that the SmtpClient is hard to Mock?

File: src/tests/Equinor.ProCoSys.IPO.Command.Tests/EventHandlers/PostSaveEvents/IpoCompletedEventHandlerTests.cs [hunk 14]
```diff
@@ -15,29 +20,63 @@ public class IpoCompletedEventHandlerTests
...
             _topicClient.Verify(t => t.SendAsync(It.IsAny<Message>()), Times.Once());
         }
+
+        [TestMethod]
+        public async Task Handle_ShouldSendEmail()
+        {
+            // Arrange
+            var objectGuid = Guid.NewGuid();
+            var plant = "PCS$TestPlant";
+            var commpkgs = new List<CommPkg>()
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.IPO.Command.Tests/EventHandlers/PostSaveEvents/IpoCompletedEventHandlerTests.cs:62]: remove () at end-of-line

File: src/tests/Equinor.ProCoSys.IPO.Command.Tests/EventHandlers/PostSaveEvents/IpoAcceptedEventHandlerTests.cs [hunk 15]
```diff
@@ -30,7 +30,7 @@ public async Task Handle_ShouldSendBusTopic()
         {
             // Arrange
             var objectGuid = Guid.NewGuid();
-            var plant = "TestPlant";
+            var plant = "PCS$TestPlant";
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.IPO.Command.Tests/EventHandlers/PostSaveEvents/IpoAcceptedEventHandlerTests.cs:33]: undo

File: src/tests/Equinor.ProCoSys.IPO.Command.Tests/EventHandlers/PostSaveEvents/IpoCanceledEventHandlerTests.cs [hunk 16]
```diff
@@ -31,7 +31,7 @@ public async Task Handle_ShouldSendBusTopic()
         {
             // Arrange
             var objectGuid = Guid.NewGuid();
-            var plant = "TestPlant";
+            var plant = "PCS$TestPlant";
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.IPO.Command.Tests/EventHandlers/PostSaveEvents/IpoCanceledEventHandlerTests.cs:34]: undo

File: src/tests/Equinor.ProCoSys.IPO.Command.Tests/EventHandlers/PostSaveEvents/IpoCompletedEventHandlerTests.cs [hunk 17]
```diff
@@ -15,29 +20,63 @@ public class IpoCompletedEventHandlerTests
...
         }
 
         [TestMethod]
         public async Task Handle_ShouldSendBusTopic()
         {
             // Arrange
             var objectGuid = Guid.NewGuid();
-            var plant = "TestPlant";
-            var ipoCompletedEvent = new IpoCompletedEvent(plant, objectGuid);
+            var plant = "PCS$TestPlant";
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.IPO.Command.Tests/EventHandlers/PostSaveEvents/IpoCompletedEventHandlerTests.cs:45]: undo

File: src/tests/Equinor.ProCoSys.IPO.Command.Tests/EventHandlers/PostSaveEvents/IpoUnAcceptedEventHandlerTests.cs [hunk 18]
```diff
@@ -30,7 +30,7 @@ public async Task Handle_ShouldSendBusTopic()
         {
             // Arrange
             var objectGuid = Guid.NewGuid();
-            var plant = "TestPlant";
+            var plant = "PCS$TestPlant";
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.IPO.Command.Tests/EventHandlers/PostSaveEvents/IpoUnAcceptedEventHandlerTests.cs:33]: undo

**Review by janid1967 — APPROVED** [review]
Approved with suggestions

File: src/Equinor.ProCoSys.IPO.Email/EmailService.cs (hunk 8 above)
Comment (janhagevold) [reply src/Equinor.ProCoSys.IPO.Email/EmailService.cs:13]: The conclusion was that adding unit tests to SmtpClient wasa more likely to introduce errors on our side. Ref. Henning

**Pull request merged**